skimage/transform/_warps.py

```diff
@@ -717,19 +717,33 @@ def _clip_warp_output(input_image, output_image, mode, cval, clip):
 
     """
     if clip:
-        min_val = input_image.min()
-        max_val = input_image.max()
-
-        preserve_cval = (mode == 'constant' and not
-                         (min_val <= cval <= max_val))
-
-        if preserve_cval:
-            cval_mask = output_image == cval
-
-        np.clip(output_image, min_val, max_val, out=output_image)
-
-        if preserve_cval:
-            output_image[cval_mask] = cval
+        min_val = np.min(input_image)
+        if np.isnan(min_val):
+            # NaNs detected, use NaN-safe min/max
+            min_func = np.nanmin
+            max_func = np.nanmax
+            min_val = min_func(input_image)
+        else:
+            min_func = np.min
+            max_func = np.max
+        max_val = max_func(input_image)
+
+        # Check if cval has been used such that it expands the effective input
+        # range
+        preserve_cval = (mode == 'constant'
+                         and not min_val <= cval <= max_val
+                         and min_func(output_image) <= cval <= max_func(output_image))
+
+        # Otherwise, set cval within the input range for clipping purposes
+        if not preserve_cval:
+            cval = min_val
+
+        # If cval is the default value, cast it to the same type as min_val
+        if cval == 0.:
+            cval = min_val.dtype.type(cval)
+
+        np.clip(output_image, min(min_val, cval), max(max_val, cval),
+                out=output_image)
```

Comment:
```suggestion
        # Otherwise, set cval within the input range for clipping purposes
        if preserve_cval:
            # cast cval to the same dtype as the input image
            cval = input_image.dtype.type(cval)
            min_val = min(min_val, cval)
            max_val = max(max_val, cval)

        np.clip(output_image, min_val, max_val, out=output_image)
```

Reply:
Refactoring slightly makes it a bit clearer to me (we don't need to do any of the cval manipulations here unless `preserve_cval` is True)

Reply:
Sure, done